### ISSUE TITLE
Treat enums as a list of constants, not dictionaries

### ISSUE THREAD

**Issue by RichardR01** [godotengine/godot-proposals#580]
Describe the project you are working on:
A brick breaker clone

Describe the problem or limitation you are having in your project:
Misunderstanding enumerations caused a strange issue with a function that converts a enumeration value to a string where the item that is being checked turned out to be a string instead of a constant.

Describe the feature / enhancement and how it helps to overcome the problem or limitation:
Because I was iterating through an enumeration "for somevar in enumVariable" somevar should have been a constant from the list in the enumeration. It turned out to be of type String. I thought it was a bug with Godot. Talking about this in Discord, they said enumerations are dictionaries. To prevent this from happening in the future (or to any other unsuspecting person), please change enumerations to be exactly what they should be "a list of constants", not a dictionary.

Here is my code and how I stumbled upon it. The match statement in bonusTypesToStr would not execute the code below the matched item even though there was a match from what is shown in the debugger. It did not show me that it was a string in the debugger (untrained eye). I interated through the enum in choseBonuses assuming that each item returned in the bonus variable would be of type int, not a string (because you know, that's how enumerations supposed to work?)

In a singleton globaL script called Global:
```
enum BonusTypes {
	NONE = -1,
	CATCH,
	BALL_SPLIT,
	PADDLE_GROW,
	BALL_SPEED,
	BONUS_SCORE,
	LIFE,
	LASERS,
	MISSILES,
	WARP
}

func rndBonus():
	var bonusChanceTable = [
		100, # CATCH
		95, # BALL_SPLIT,
		90, # PADDLE_GROW
		80, # BALL_SPEED,
		70, # BONUS_SCORE,
		50, # LIFE,
		40, # LASERS,
		20, # MISSILES,
		5, # WARP
	]
	var bcTotals = []
	var bcTotal = 0
	for b in bonusChanceTable:
		bcTotal += b
		bcTotals.append(bcTotal)
	var rnd = int(round(randf() * bcTotal))
	bcTotals.sort()
	var r = bcTotals.bsearch(rnd)
	bcTotals.clear()
	return r

func choseBonuses(numChosenBonuses):
	var maxBonusAmount = int(numChosenBonuses / (BonusTypes.size() - 1)) + 1
	for bonus in BonusTypes:
		for i in range(maxBonusAmount):
			chosenBonuses.append(bonus)
	chosenBonuses.shuffle()

func bonusTypeToStr(bonus):
	var r = ""
	match bonus:
		BonusTypes.NONE:
			r = "None"
		BonusTypes.CATCH:
			r = "Catch"
		BonusTypes.BALL_SPLIT:
			r = "Ball Split"
		BonusTypes.PADDLE_GROW:
			r = "Paddle Grow"
		BonusTypes.BALL_SPEED:
			r = "Ball Speed"
		BonusTypes.BONUS_SCORE:
			r = "Bonus Score"
		BonusTypes.LIFE:
			r = "Extra Life"
		BonusTypes.LASERS:
			r = "Lasers Type 1"
		BonusTypes.MISSILES:
			r = "Missiles"
		BonusTypes.WARP:
			r = "Warping..."
	return r
```
In another script
```
	Global.choseBonuses(numBricks)
	for i in range(Global.chosenBonuses.size()):
		print(Global.bonusTypeToStr(Global.chosenBonuses[i]))
```

**Comment by KoBeWi** [comment]
This is already planned for GDScript changes in Godot 4.0 or 4.1.

**Comment by vnen** [comment]
There's not really much way around this. To keep enums accessible at runtime they must be in a dictionary, since there's not any other structure available to perform the same behavior. I believe this is just a matter of documentation, because when you iterate a dictionary you get its keys, not its values.

Many languages do not allow you access to enums like that, much less iterating through them, so that is a bonus of using dictionaries.

**Comment by KoBeWi** [comment]
Implemented in 4.0

**Comment by Binozo** [comment]
> Implemented in 4.0

https://www.reddit.com/r/ProgrammerHumor/comments/u506ft/impossi_implemented/?utm_source=share&utm_medium=web2x&context=3

**Comment by KoBeWi** [comment]
Funny, because it was reverted due to many bugs it caused.
